### DISCUSSION TITLE
How are we supposed to use useInfiniteQuery with REST APis that use self/next/prev urls

### DISCUSSION THREAD

**Discussion by anilanar** [TanStack/query#6755] [Q&A]
For example, see Confluence pagination: https://developer.atlassian.com/server/confluence/pagination-in-the-rest-api/

So pageParams are urls to other pages. Therefore there is no "initialPageParam". But with v5, initialPageParam is required.

We can work around it by either passing empty string as `initialPageParam` (and in queryFn, check if pageParam is empty string or not) or by passing the original url as `initialPageParam`. But that's far from being elegant. Why make it required?

**Comment by TkDodo** [comment]
`initialPageParam` is required because a) we can infer the type for `pageParam` from it and b) we used to just pass `undefined` for the first page, but that causes trouble in serialization because `undefined` is not serializable.

You can also set your `initialPageParam` to the url itself and then always use it:

```
useInfiniteQuery({
  queryKey: ['thing'],
  queryFn: ({ pageParam }) => fetch(pageParam),
  initialPageParam: '/thing',
  getNextPageParam: (lastPage) => lastPage._links.next,
})
```

**Reply by syz51** [reply]
hey qq. So in the case of `initialData` being set, the `initialPageParam` is never used, right? Because the initial request will never be sent. In this case, the `initialPageParam` is only used to infer the type for `pageParam`.
For example:
```
useInfiniteQuery({
    queryKey: ["videos"],
    queryFn: async ({ pageParam }) => {
      return await fetchVideos(pageParam);
    },
    getNextPageParam: (lastPage: Video[], _) =>
      lastPage.length === 0
        ? undefined
        : {
            lastStreamingDate: lastPage[lastPage.length - 1].streamingDate,
            lastCode: lastPage[lastPage.length - 1].code,
          },
    initialPageParam: {
      lastStreamingDate: "",
      lastCode: "",
    },
    initialData: { pages: [initialVideos], pageParams: [] },
  });
```

**Reply by TkDodo** [reply]
well if you refetch the query and this is the first page / only page in the cache, we will use the `initialPageParam` to refetch that page.
